lib/src/exe/try_join_all_buffered.rs

```diff
@@ -0,0 +1,79 @@
+use futures::{
+	future::IntoFuture, ready, stream::FuturesOrdered, TryFuture, TryFutureExt, TryStream,
+};
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::mem;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+pin_project! {
+	/// Future for the [`try_join_all_buffered`] function.
+	#[must_use = "futures do nothing unless you `.await` or poll them"]
+	pub struct TryJoinAllBuffered<F, I>
+	where
+		F: TryFuture,
+		I: Iterator<Item = F>,
+	{
+		input: I,
+		#[pin]
+		active: FuturesOrdered<IntoFuture<F>>,
+		output: Vec<F::Ok>,
+	}
+}
+
+/// Creates a future which represents either an in-order collection of the
+/// results of the futures given or a (fail-fast) error.
+///
+/// Only a limited number of futures are driven at a time.
+pub fn try_join_all_buffered<I>(iter: I) -> TryJoinAllBuffered<I::Item, I::IntoIter>
+where
+	I: IntoIterator,
+	I::Item: TryFuture,
+{
+	let mut input = iter.into_iter();
+	let mut active = FuturesOrdered::new();
+
+	#[cfg(target_arch = "wasm32")]
+	const LIMIT: usize = 1;
+
+	#[cfg(not(target_arch = "wasm32"))]
+	const LIMIT: usize = crate::cnf::MAX_CONCURRENT_TASKS;
+
+	while active.len() < LIMIT {
+		if let Some(next) = input.next() {
+			active.push_back(TryFutureExt::into_future(next));
+		} else {
+			break;
+		}
+	}
+
+	TryJoinAllBuffered {
+		input,
+		active,
+		output: Vec::new(),
```

Comment:
Would it make sense to try and pre-allocate this using `Iterator::size_hint()`? That should help reduce allocations in some cases.

Reply:
Good catch!